Курсовая Ежедневник. На проверку.

KaerLaende opened the pull request.

KaerLaende committed:
Курсовая Ежедневник. На проверку.

KaerLaende committed:
Курсовая Ежедневник. На проверку. Ура!!! смог правильно реализовать добавление задач на новые даты автоматически, в зависимости от типа повторяемости!

KaerLaende committed:
Курсовая Ежедневник. На проверку. Ура!!! После просмотра урока :Коллекции. Часть 3. Нашел решения не упорядоченных дат при выводе всего Map. Теперь появилась какая то упорядоченность!)))

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,15 +1,196 @@
+import java.util.Scanner;
 
 public class Main {
+    public static DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
```
если не используется вне класса, то зачем public? инкапсуляцию никто не отменял

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,15 +1,196 @@
 
 public class Main {
+    public static DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+    public static DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm");
```
если не используется вне класса, то зачем public? инкапсуляцию никто не отменял

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,15 +1,196 @@
+            String title = getTitle(scanner);
+            String description = getDescription(scanner);
+            int taskType = getTaskType(scanner);
+            Class<? extends Type> type = taskType == 1 ? WORK.class : PERSONAL.class;
```
а это что такое? тип задачи должен быть enum (см. диаграмму классов в задании)

KaerLaende replied:
Да я подумал что в этой диаграмме значек что рядом с другими {} классами, он же рядом и с {} WORK и подумал что как то они странно предложили сделать это тоже классами, но коль просили решил мучатся с классами, но все равно получилось!)

dmitry-bizin replied:
не, всё гораздо проще) и, кстати, курсовая по максимуму включает темы 2 курса, поэтому обрати на это внимание (дженериков только нет своих, в коллекциях только)) )

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,15 +1,196 @@
+    private static void printAllTasksForDay(Scanner scanner) {
+        System.out.print("Введите дату для получения событий и задач в формате ггг-мм-дд: ");
+        String date = scanner.next();
+        LocalDate localDate = LocalDate.parse(date, dateFormatter);
```
нет обработки некорректно введенной даты

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,15 +1,196 @@
+
+    private static void printAllDeletedTasks(){
+        System.out.println("Завершенные события:");
+        TaskService.getRemovedTasks();
```
почему метод называется get, но ничего не возвращает?

dmitry-bizin commented on src/Task/DailyTask.java:
```diff
@@ -1,14 +1,41 @@
 import java.time.LocalDateTime;
+import java.time.LocalTime;
 
 public class DailyTask<T extends Type> extends Task {
```
не, здесь нужно было сделать немного не так, а как показано на диаграмме классов (см. задание)

dmitry-bizin commented on src/Task/DailyTask.java:
```diff
@@ -1,14 +1,41 @@
+            TaskService.add(cloneTask(this,i));
+        }
     }
+    public static DailyTask cloneTask(Task task, int i) {
```
это всё не нужно

KaerLaende replied:
Да это вообще была вершина мысли!!=))Я понимал что наверно это не обязательно, но продолжаю считать, что так все таки правильнее. если есть задача ежедневная и  еженедельная и я запрошу все задачи в тот день где они обе присутствуют?

dmitry-bizin replied:
я в фидбеке написал, как и что должно быть)

dmitry-bizin replied:
там при помощи метода это можно будет определить всё

dmitry-bizin commented on src/Task/DailyTask.java:
```diff
@@ -1,14 +1,41 @@
+    public DailyTask(String title, Class type, LocalDate date, LocalTime time, String description) {
+        super(title, type, date, time, description);
+        for (int i = 1; i < 3; i++) {
+            TaskService.add(cloneTask(this,i));
```
это всё не нужно

dmitry-bizin replied:
ты можешь оставить клонов, но не нужно добавлять задачу в конструкторе - это нарушает принцип единственной ответственности SOLID - у тебя конструктор занимается ещё другим делом, кроме как созданием объекта

dmitry-bizin commented on src/Task/DailyTask.java:
```diff
@@ -1,14 +1,41 @@
 
 
+    }
+    public LocalDate getNextDay(Task t) {
```
здесь по-другому нужно сделать (прикреплю пример в фидбеке)

dmitry-bizin commented on src/Task/MonthlyTask.java:
```diff
@@ -1,11 +1,38 @@
+            TaskService.add(cloneTask(this,i));
+        }
+    }
+    public static MonthlyTask cloneTask(Task task, int i) {
```
это всё не нужно

dmitry-bizin commented on src/Task/MonthlyTask.java:
```diff
@@ -1,11 +1,38 @@
+    public MonthlyTask(String title, Class type,  LocalDate date, LocalTime time, String description) {
+        super(title, type, date, time, description);
+        for (int i = 1; i < 2; i++) {
+            TaskService.add(cloneTask(this,i));
```
это всё не нужно

dmitry-bizin replied:
ты можешь оставить клонов, но не нужно добавлять задачу в конструкторе - это нарушает принцип единственной ответственности SOLID - у тебя конструктор занимается ещё другим делом, кроме как созданием объекта

dmitry-bizin commented on src/Task/Task.java:
```diff
@@ -1,81 +1,115 @@
 
-public class Task <T extends Type>{
-    private static int idGanerator =1;
+public class Task <T extends Type> implements Cloneable{
```
класс не должен быть дженериков, в этом задании не будет дженериков, Cloneable тоже не должен быть

dmitry-bizin replied:
зачем тебе дженерика-параметр, если он не используется?

dmitry-bizin commented on src/Task/Task.java:
```diff
@@ -1,81 +1,115 @@
-    LocalDateTime dateTime;
-    private String description;
 
+    private LocalDate date;
```
зачем дата и дата и время? должно быть только дата и время (см. диаграмму классов)

dmitry-bizin replied:
вместо 3 полей: date, dateTime и time должно быть 1 поле - dateTime (см. иерархию классов)

dmitry-bizin commented on src/Task/WeeklyTask.java:
```diff
@@ -1,11 +1,40 @@
+        }
+    }
+    public static WeeklyTask cloneTask(Task task, int i) {
+        WeeklyTask newTask= null;
```
это всё не нужно

dmitry-bizin commented on src/Task/WeeklyTask.java:
```diff
@@ -1,11 +1,40 @@
+    public WeeklyTask(String title, Class type,  LocalDate date, LocalTime time, String description) {
+        super(title, type, date, time, description);
+        for (int i = 1; i < 3; i++) {
+            TaskService.add(cloneTask(this,i));
```
это всё не нужно

dmitry-bizin replied:
ты можешь оставить клонов, но не нужно добавлять задачу в конструкторе - это нарушает принцип единственной ответственности SOLID - у тебя конструктор занимается ещё другим делом, кроме как созданием объекта

dmitry-bizin commented on src/Task/YearlyTask.java:
```diff
@@ -1,11 +1,41 @@
+    public static YearlyTask cloneTask(Task task, int i) {
+        YearlyTask newTask= null;
+        try {
+             newTask = (YearlyTask) task.clone();
```
это всё не нужно

dmitry-bizin commented on src/Task/YearlyTask.java:
```diff
@@ -1,11 +1,41 @@
+        super(title, type,  date, time, description);
+
+        for (int i = 1; i < 100; i++) {// простите рука не поднялась изменить количество повторов на меньшее...
+            TaskService.add(cloneTask(this,i));
```
это всё не нужно

dmitry-bizin commented on src/Task/YearlyTask.java:
```diff
@@ -1,11 +1,41 @@
+    public YearlyTask(String title, Class type, LocalDate date, LocalTime time, String description) {
+        super(title, type,  date, time, description);
+
+        for (int i = 1; i < 100; i++) {// простите рука не поднялась изменить количество повторов на меньшее...
```
это и не нужно)

dmitry-bizin replied:
ты можешь оставить клонов, но не нужно добавлять задачу в конструкторе - это нарушает принцип единственной ответственности SOLID - у тебя конструктор занимается ещё другим делом, кроме как созданием объекта

dmitry-bizin commented on src/TaskService/IncorrectArgumentException.java:
```diff
@@ -0,0 +1,27 @@
+package TaskService;
+
+public class IncorrectArgumentException extends RuntimeException{
```
исключение должно быть проверяемым

dmitry-bizin commented on src/TaskService/TaskNotFoundException.java:
```diff
@@ -1,4 +1,27 @@
 package TaskService;
 
-public class TaskNotFoundException {
+public class TaskNotFoundException extends RuntimeException{
```
исключение должно быть проверяемым

dmitry-bizin commented on src/TaskService/TaskService.java:
```diff
@@ -1,18 +1,62 @@
-//    public Collection<Task> getAllByDate(LocalDate ld){
-//         }
+    private static Map<Integer, Task> taskMap = new LinkedHashMap<Integer, Task>();
+    private static ArrayList<Task> removedTasks;
```
List

dmitry-bizin commented on src/TaskService/TaskService.java:
```diff
@@ -1,18 +1,62 @@
+    private static ArrayList<Task> removedTasks;
+
+    public static Map<Integer, Task> getTaskMap() {
+        return taskMap;
```
нет инкапсуляции

dmitry-bizin commented on src/TaskService/TaskService.java:
```diff
@@ -1,18 +1,62 @@
+        return taskMap;
+    }
+
+    public static void getRemovedTasks() {
```
get? а почему void?

dmitry-bizin commented on src/TaskService/TaskService.java:
```diff
@@ -1,18 +1,62 @@
+    }
+    public static void removeTask(int id) throws TaskNotFoundException{
+        if (taskMap.containsKey(id)) {
+               removedTasks.add(taskMap.remove(id));//здесь ошибка!? Не могу передать taskMap.getValue(get(id))?
```
не очень понял

KaerLaende replied:
Я не совсем понял как правильно добавить в Список Task'ов позицию (ключ, значение).. и // показывал что думаю тут как то надо выделить из Мар именно вэлью (то есть содержания = Task)?

dmitry-bizin replied:
вот как у тебя сейчас написано - так и должно быть. Ты удаляешь из мапы и сразу кладёшь в коллекцию

KaerLaende replied:
пффф это не мой вариант - слишком просто!!! Мне надо как с клонами, что бы часов 6 потратить на написание одного метода и что бы при этом получилось только через день! )

dmitry-bizin commented on src/TaskService/TaskService.java:
```diff
@@ -1,18 +1,62 @@
+    public static void removeTask(int id) throws TaskNotFoundException{
+        if (taskMap.containsKey(id)) {
+               removedTasks.add(taskMap.remove(id));//здесь ошибка!? Не могу передать taskMap.getValue(get(id))?
+               taskMap.remove(id);
```
это не нужно, зачем второй раз удаляешь? ты же выше удалил уже

KaerLaende replied:
Контрольный выстрел=))) Что б наверняка!!! :)

dmitry-bizin commented on src/TaskService/TaskService.java:
```diff
@@ -1,18 +1,62 @@
+    public static void getAllByDate(LocalDate ld) {
+        taskMap.forEach((a,b)-> System.out.print((Objects.equals(b.getDate(), ld)?b:"")));
+    // несколько часов писал этот метод, не смог сам разобраться как это сделать если..
+    // ..у нас LocalDateTime в переменной Задачи, а нас просят здесь указать LocalDate?
```
см. пример в фидбеке

KaerLaende replied:
Поэтому и сделал через Локал дате и Локал тайм, а дальше сравнивать легче: Локал дате из параметра Таск и Локал дате в других методах, где дата не нужна

dmitry-bizin replied:
Из LocalDateTime можно получить LocalDate

dmitry-bizin commented on src/TypeOfTask/PERSONAL.java:
```diff
@@ -1,8 +1,6 @@
 package TypeOfTask;
 
 public class PERSONAL extends Type{
```
это не нужно

dmitry-bizin commented on src/TypeOfTask/WORK.java:
```diff
@@ -1,8 +1,5 @@
 package TypeOfTask;
 
 public class WORK extends Type{
```
это не нужно

dmitry-bizin commented on src/TypeOfTask/Type.java:
```diff
@@ -1,11 +1,16 @@
 package TypeOfTask;
 
-public class Type {
+public class Type implements Cloneable {
```
это должен быть enum и не нужен Cloneable

KaerLaende committed:
Исправления:
1 Переделал Type на enum;
2 Исправил Exeption;
3 добавил проверку даты при использовании сканера;
4 некоторые другие корректировки...
5 по братски! Я же страдал... можно клонов оставить - это же крутяк!? :)

dmitry-bizin commented on src/Task/Task.java:
```diff
@@ -24,8 +23,12 @@ public class Task <T extends Type> implements Cloneable{
+        try {
+            setTitle(title);
+        } catch (IncorrectArgumentException e) {
+            throw new RuntimeException(e);
```
не надо ловить исключение, смысл тогда его выбрасывать? пробрасывай своё исключение из конструктора

dmitry-bizin commented on src/TaskService/TaskService.java:
```diff
@@ -47,16 +47,13 @@ public static Task getTask(int id) throws IncorrectArgumentException {
             }
         }
     public static void getAllByDate(LocalDate ld) {
         taskMap.forEach((a,b)-> System.out.print((Objects.equals(b.getDate(), ld)?b:"")));
```
а где здесь исправления? тебе надо переделать иерархию классов-задач, чтобы каждый наследник реализовывал по-своему метод проверки задачи на дату, здесь надо использовать этот метод. Сейчас у тебя не работает метод получения задач на дату
